runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java

```diff
@@ -359,13 +533,100 @@ public void finishBundle() {
         emitResults();
       } catch (Exception e) {
         throw new RuntimeException("Failed to finish remote bundle", e);
+      } finally {
+        remoteBundle = null;
+      }
+      if (bundleFinishedCallback != null) {
+        bundleFinishedCallback.run();
+        bundleFinishedCallback = null;
       }
     }
 
+    /** Key for timer which has not been registered yet. */
+    Object getTimerKeyForRegistration() {
+      return keyForTimerToBeSet;
+    }
+
+    /** Key for timer which is about to be fired. */
+    void setTimerKeyForFire(Object key) {
+      this.keyForTimerToBeFired = key;
+    }
+
+    boolean isBundleInProgress() {
+      return remoteBundle != null;
+    }
+
+    void setBundleFinishedCallback(Runnable callback) {
+      this.bundleFinishedCallback = callback;
+    }
+
     private void emitResults() {
       KV<String, OutputT> result;
       while ((result = outputQueue.poll()) != null) {
-        outputManager.output(outputMap.get(result.getKey()), (WindowedValue) result.getValue());
+        final String outputPCollectionId = Preconditions.checkNotNull(result.getKey());
+        TupleTag<?> tag = outputMap.get(outputPCollectionId);
+        WindowedValue windowedValue =
+            Preconditions.checkNotNull(
+                (WindowedValue) result.getValue(),
+                "Received a null value from the SDK harness for %s",
+                outputPCollectionId);
+        if (tag != null) {
+          // process regular elements
+          outputManager.output(tag, windowedValue);
+        } else {
+          final String timerCollectionId = extractTimerPCollectionId(outputPCollectionId);
+          TimerSpec timerSpec = extractTimerSpec(timerCollectionId);
+          Timer timer =
+              Preconditions.checkNotNull(
+                  (Timer) ((KV) windowedValue.getValue()).getValue(),
+                  "Received null Timer from SDK harness: %s",
+                  windowedValue);
+          LOG.debug("Timer received: {} {}", outputPCollectionId, timer);
+          for (Object window : windowedValue.getWindows()) {
+            StateNamespace namespace = StateNamespaces.window(windowCoder, (BoundedWindow) window);
+            TimerInternals.TimerData timerData =
+                TimerInternals.TimerData.of(
+                    timerCollectionId, namespace, timer.getTimestamp(), timerSpec.getTimeDomain());
+            setTimer(windowedValue, timerData);
+          }
+        }
+      }
+    }
+
+    private String extractTimerPCollectionId(String outputPCollectionId) {
+      return stageBundleFactory
+          .getProcessBundleDescriptor()
+          .getProcessBundleDescriptor()
+          .getPcollectionsMap()
+          .get(outputPCollectionId)
+          .getUniqueName();
```

Comment:
The name here is somehow not stable anymore after #6986.

Reply:
@robertwb Reverted #6986 and it works again.

Reply:
For example, I have `ParDo.timer.foo.out:0` which resolves to `ParDo.timer.output` instead of `ParDo.timer.foo`.

Reply:
Unless I'm mistaken, extractTimerPCollectionId should not be needed at all anymore, as timerReferenceMap is indexed by outputPCollectionId itself.

Reply:
It is still necessary. The timerReferenceMap gives me `ParDo.timer.foo` but while processing the timer elements, I get `ParDo.timer.foo.out:0`. The issue seems to be that the output collection ids are not set correctly.

Reply:
>Unless I'm mistaken, extractTimerPCollectionId should not be needed at all anymore, as timerReferenceMap is indexed by outputPCollectionId itself.

It is still necessary. The timerReferenceMap gives me `ParDo.timer.foo` for the output name of timer "foo", when it is `ParDo.timer.foo.out:0` at runtime.  I believe there lies the problem.